sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/statefulOperators.scala

```diff
@@ -320,11 +322,22 @@ trait StateStoreWriter
    * the driver after this SparkPlan has been executed and metrics have been updated.
    */
   def getProgress(): StateOperatorProgress = {
+    // Still publish the instance metrics that are marked as unpopulated,
+    // as they represent state store instances that never uploaded a snapshot.
+    val instanceMetrics = stateStoreInstanceMetrics.map { entry =>
+      entry._1 -> (if (longMetric(entry._1).isZero) -1L else longMetric(entry._1).value)
+    }
+    // Only keep the smallest N instance metrics to report to driver.
```

Comment:
This is not scalable. For the last uploaded snapshot, the top N is the smallest N because smaller value = bigger lag. But what if we want to have the metric for memory usage and want to produce the top N excessive memory usage?

Whenever we deal with interface, we need to consider the interface as "general" one and think through possible derivations. Since we only allow a numeric value for that metric, probably describing sort order in the interface would be fine. (Otherwise we may have to enforce implementing Comparable for the derivation of the interface.)

Reply:
Once we reflect this, we'd like to generalize the comment as well e.g. pick N with specified order.

Reply:
This makes sense, I've just added some configuration options to the instance metrics class to help generalize this to some more custom metrics that we may want to add in the future.